File: almoosa_customization/almoosa_customization/report/summary_sales_by_brand_and_group/summary_sales_by_brand_and_group.js
// Copyright (c) 2026, Printechs and contributors
// For license information, please see license.txt

frappe.query_reports["Summary Sales By Brand And Group"] = {
	"filters": [
		 {
            "fieldname": "from_datetime",
            "label": "From Date & Time",
            "fieldtype": "Datetime",
            "default": frappe.datetime.month_start() + " 00:00:01"
        },

        {
            "fieldname": "to_datetime",
            "label": "To Date & Time",
            "fieldtype": "Datetime",
            "default": frappe.datetime.month_end() + " 23:59:59"
        },
		{
			"fieldname": "warehouse",
			"label": "Warehouse",
			"fieldtype": "MultiSelectList",
			"options": "Warehouse",
			"get_data": function(txt) {
				return frappe.db.get_link_options("Warehouse", txt);
			}
		},
        {
			"fieldname": "brand",
			"label": "Brand",
			"fieldtype": "MultiSelectList",
			"options": "Brand",
			"get_data": function(txt) {
				return frappe.db.get_link_options("Brand", txt);
			}
		}
    ],
	
	"formatter": function (value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        if (!data) {
            return value;
        }

        // ✅ Bold TOTAL column (all rows)
        if (column.fieldname === "total") {
            value = `<span style="font-weight:600;">${value}</span>`;
        }

        // ✅ Bold GRAND TOTAL row
        if (data.item_group === "Grand Total") {
            value = `<span style="font-weight:700;">${value}</span>`;
        }

        return value;
}

};

File: almoosa_customization/almoosa_customization/report/summary_sales_by_brand_and_group/summary_sales_by_brand_and_group.py
# Copyright (c) 2026, Printechs and contributors
# For license information, please see license.txt

import frappe
from collections import defaultdict

def execute(filters=None):

	filters = filters or {}
	conditions = "hd.docstatus = 1"
	values = {}

	if filters.get("from_datetime") and filters.get("to_datetime"):
		conditions += """
        AND TIMESTAMP(hd.posting_date, hd.posting_time)
        BETWEEN %(from_datetime)s AND %(to_datetime)s
    	"""
		values["from_datetime"] = filters.get("from_datetime")
		values["to_datetime"] = filters.get("to_datetime")
    
	if filters.get("warehouse"):
		conditions += " AND it.warehouse IN %(warehouse)s"
		values["warehouse"] = tuple(filters.get("warehouse"))

	if filters.get("brand"):
		conditions += " AND it.brand IN %(brand)s"
		values["brand"] = tuple(filters.get("brand"))

	data = frappe.db.sql(f"""
    SELECT
        LEFT(it.warehouse,3) warehouse,
        SUBSTRING_INDEX(it.item_group, '.', -1) AS item_group,
        it.brand,
        SUM(it.qty) AS qty
    FROM `tabPOS Invoice Item` it
    INNER JOIN `tabPOS Invoice` hd
        ON hd.name = it.parent
    WHERE {conditions}
    GROUP BY
        it.warehouse,
        item_group,
        it.brand
    ORDER BY
        it.warehouse, item_group
""", values, as_dict=True)

    # -------------------------
    # Collect Dynamic Brands
    # -------------------------

	#frappe.msgprint(str(values))
	brands = sorted(list({d.brand for d in data if d.brand}))

	columns = [
        {"label": "WH", "fieldname": "warehouse", "fieldtype": "Data", "width": 60},
        {"label": "Group", "fieldname": "item_group", "fieldtype": "Data", "width": 100},
    ]

	for brand in brands:
		columns.append({
            "label": brand,
            "fieldname": frappe.scrub(brand),
            "fieldtype": "Int",
            "width": 70
        })

	columns.append({
        "label": "Total",
        "fieldname": "total",
        "fieldtype": "Int",
        "width": 100
    })

    # -------------------------
    # Organize Data
    # -------------------------
	warehouse_data = defaultdict(lambda: defaultdict(dict))
	warehouse_totals = defaultdict(lambda: defaultdict(float))
	grand_totals = defaultdict(float)

	for row in data:
		warehouse = row.warehouse
		group = row.item_group
		brand = row.brand
		qty = row.qty or 0

		warehouse_data[warehouse][group][brand] = qty
		warehouse_totals[warehouse][brand] += qty
		grand_totals[brand] += qty
		warehouse_totals[warehouse]["total"] += qty
		grand_totals["total"] += qty

    # -------------------------
    # Build Final Rows
    # -------------------------
	result = []

	for warehouse in sorted(warehouse_data.keys()):
		first_row = True

		for group in sorted(warehouse_data[warehouse].keys()):
			row = {
                "warehouse": warehouse if first_row else "",
                "item_group": group
            }

			total = 0
			for brand in brands:
				qty = warehouse_data[warehouse][group].get(brand, 0)
				row[frappe.scrub(brand)] = qty
				total += qty

			row["total"] = total
			result.append(row)
			first_row = False

        # Warehouse Total Row
		total_row = {
            "warehouse": "",
            "item_group": "Warehouse Total"
        }

		for brand in brands:
			total_row[frappe.scrub(brand)] = warehouse_totals[warehouse][brand]

		total_row["total"] = warehouse_totals[warehouse]["total"]
		result.append(total_row)

    # Grand Total Row
	grand_row = {
        "warehouse": "",
        "item_group": "Grand Total"
    }

	for brand in brands:
		grand_row[frappe.scrub(brand)] = grand_totals[brand]

	grand_row["total"] = grand_totals["total"]
	result.append(grand_row)

	return columns, result
